doschema/jsvalidation.py

```diff
@@ -0,0 +1,232 @@
+#
+# DoSchema is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+
+import jsonschema
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver: Resolver used to retrieve referenced schemas. If not
+                         provided it will use default.
+
+        """
+        self.ignore_index = ignore_index
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        self.traverse(schema, ())
+
+        self.uri = None
+        self.resolver = None
+
+    def traverse(self, curr_schema, curr_field, check_dependencies=True):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        :param check_dependencies:
+        """
+        self.check_special(curr_schema, curr_field)
+
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        if curr_schema.get('type') == 'object':
+            if 'properties' in curr_schema:
+                props = curr_schema['properties']
+
+                for prop in props:
+                    if isinstance(props[prop], dict):
+                        while '$ref' in props[prop]:
+                            path = props[prop].get('$ref')
+                            props[prop] = self.resolver.resolve(path)[1]
```

Comment:
Modifying a collection while iterating on it is not a very good design. Here it works but you can't know in advance how the code will be refactored later. Thus I would recommend having a separate variable for the value instead of using `props[prop]`. It would also be better to consider the provided schema as *read only* as it will make debug easier.
You can even use [six.iteritems](https://pythonhosted.org/six/#six.iteritems) to iterate over the items of the dict in a consistent manner in Python 2 and 3.